src/webgpu/util/texture/texelData.spec.ts

```diff
@@ -0,0 +1,377 @@
+export const description = 'Test helpers for texel data produce the expected data in the shader';
+
+import { params, poptions } from '../../../common/framework/params_builder.js';
+import { makeTestGroup } from '../../../common/framework/test_group.js';
+import { unreachable, assert } from '../../../common/framework/util/util.js';
+import {
+  kEncodableTextureFormats,
+  kEncodableTextureFormatInfo,
+  EncodableTextureFormat,
+} from '../../capability_info.js';
+import { GPUTest } from '../../gpu_test.js';
+
+import { getTexelDataRepresentation, TexelComponent } from './texelData.js';
+
+export const g = makeTestGroup(GPUTest);
+
+let ReadbackTypedArray: Float32ArrayConstructor | Int32ArrayConstructor | Uint32ArrayConstructor;
+
+function doTest(
+  t: GPUTest,
+  format: EncodableTextureFormat,
+  componentData: { [C in TexelComponent]?: number }
+) {
+  const rep = getTexelDataRepresentation(format);
+  const texelData = rep.packData(componentData);
+  const texture = t.device.createTexture({
+    format,
+    size: [1, 1, 1],
+    usage: GPUTextureUsage.COPY_DST | GPUTextureUsage.SAMPLED,
+  });
+
+  t.device.defaultQueue.writeTexture(
+    { texture },
+    texelData,
+    {
+      bytesPerRow: texelData.byteLength,
+    },
+    [1, 1, 1]
+  );
+
+  let shaderType: 'i32' | 'u32' | 'f32';
+  switch (kEncodableTextureFormatInfo[format].componentType) {
+    case 'sint':
+      ReadbackTypedArray = Int32Array;
+      shaderType = 'i32';
+      break;
+    case 'uint':
+      ReadbackTypedArray = Uint32Array;
+      shaderType = 'u32';
+      break;
+    case 'float':
+      ReadbackTypedArray = Float32Array;
+      shaderType = 'f32';
+      break;
+    default:
+      unreachable();
+  }
+
+  const shader = `
+  [[set(0), binding(0)]] var<uniform_constant> tex : texture_2d<${shaderType}>;
+
+  [[block]] struct Output {
+    ${rep.componentOrder
+      .map((C, i) => {
+        return `[[offset(${i * 4})]] result${C} : ${shaderType};`;
+      })
+      .join('\n')}
```

Comment:
Could just return a `vec4<shaderType>`. Then you can just `output.result = textureLoad(...)` I think. In the result you can just leave the unused components untested so we're not testing that behavior here (should be a separate test).

Reply:
Maybe, but then I can't just use t.expectContents of the whole thing?

Reply:
True. Couldn't you still make this a simple vec4 and write into it selectively in main()? (Or can you not put swizzles on the left of `=` in wgsl?) Anyway, either is okay.